xarray/core/coordinates.py

```diff
@@ -341,9 +343,11 @@ class DataArrayCoordinates(Coordinates):
     dimensions and the values given by corresponding DataArray objects.
     """
 
+    _data: T_DataArray
+
     __slots__ = ("_data",)
 
-    def __init__(self, dataarray: DataArray):
+    def __init__(self, dataarray: T_DataArray):
```

Comment:
I think you need to make the class generic as well. Not sure how this works when inheriting from Coords which is a generic `Mapping[Hashable, T_DataArray]`.